File: chapter-02/exercise-02/pointer_declarations.cpp
#include <iostream>
#include <type_traits>

int main() {
    int y = 1;
    char ch = 'A';
    long int year = 2026;
    double amount = 125.50;
    int zValue = 25;
    float quantity = 3.5F;
    int date = 29;
    double yield = 8.75;

    int* yAddr = &y;
    char* chAddr = &ch;
    long int* ptYr = &year;
    double* amt = &amount;
    int* z = &zValue;
    float* qp = &quantity;
    int* datePt = &date;
    double* yldAddr = &yield;

    static_assert(std::is_same_v<decltype(yAddr), int*>);
    static_assert(std::is_same_v<decltype(chAddr), char*>);
    static_assert(std::is_same_v<decltype(ptYr), long int*>);
    static_assert(std::is_same_v<decltype(amt), double*>);
    static_assert(std::is_same_v<decltype(z), int*>);
    static_assert(std::is_same_v<decltype(qp), float*>);
    static_assert(std::is_same_v<decltype(datePt), int*>);
    static_assert(std::is_same_v<decltype(yldAddr), double*>);

    std::cout << "yAddr -> " << *yAddr << '\n';
    std::cout << "chAddr -> " << *chAddr << '\n';
    std::cout << "ptYr -> " << *ptYr << '\n';
    std::cout << "amt -> " << *amt << '\n';
    std::cout << "z -> " << *z << '\n';
    std::cout << "qp -> " << *qp << '\n';
    std::cout << "datePt -> " << *datePt << '\n';
    std::cout << "yldAddr -> " << *yldAddr << '\n';
}

File: chapter-02/exercise-02/pointer_declarations.py
"""Show Python references corresponding to the presentation's C++ pointers."""


def main() -> None:
    # Python variables do not need typed pointer declarations.
    y_addr = 1
    ch_addr = "A"
    pt_yr = 2026
    amt = 125.50
    z = 25
    qp = 3.5
    date_pt = 29
    yld_addr = 8.75

    references = {
        "yAddr": y_addr,
        "chAddr": ch_addr,
        "ptYr": pt_yr,
        "amt": amt,
        "z": z,
        "qp": qp,
        "datePt": date_pt,
        "yldAddr": yld_addr,
    }

    assert isinstance(references["yAddr"], int)
    assert isinstance(references["chAddr"], str)
    assert isinstance(references["ptYr"], int)
    assert isinstance(references["amt"], float)
    assert isinstance(references["z"], int)
    assert isinstance(references["qp"], float)
    assert isinstance(references["datePt"], int)
    assert isinstance(references["yldAddr"], float)

    for name, value in references.items():
        print(f"{name:<8} -> {value!r}")


if __name__ == "__main__":
    main()
